test/batch_exchange_utils.js

```diff
@@ -0,0 +1,100 @@
+const MockContract = artifacts.require("MockContract")
+
+const BN = require("bn.js")
+
+const { closeAuction, getBalanceState, getWithdrawableAmount } = require("../scripts/utilities.js")
+
+const { setupGenericStableX } = require("./stablex_utils")
+
+contract("BatchExchange utils", async (accounts) => {
+  describe("getBalanceState()", async () => {
+    it("retrieves balance as in storage", async () => {
+      const batchExchange = await setupGenericStableX()
+      await closeAuction(batchExchange)
+      const erc20 = await MockContract.new()
+      await erc20.givenAnyReturnBool(true)
+
+      await batchExchange.deposit(erc20.address, 0x100)
+      await closeAuction(batchExchange)
+      // force balance update by creating new deposit
+      await batchExchange.deposit(erc20.address, 0)
+
+      const balance = await getBalanceState(accounts[0], erc20.address, batchExchange.address, web3)
+      assert.equal(balance.toString(16), "100")
+    })
+  })
+
+  describe("getWithdrawableAmount()", async () => {
+    const startingAmount = new BN("100")
+
+    const setup = async function () {
```

Comment:
I'm usually not a big fan of common setup methods in unit tests unless they provide significant value. 
They require the maintainer to context switch from the test case into another method in order to understand what state is given and tend to contain either more sub-methods (more context switches) or one-off logics which are used for one test case but not the others.

Here it seems like we should be using just a raw new EpochTokenLocker (not the batchExchange) and the small setup (depositing from a MockedToken) can happen inside each test case. This way it's easy to understand what's going on in a single method.

Reply:
I believe we should not use `EpochTokenLocker` but we should stick to `BatchExchange`.
The function `getBalanceState` hardcodes the storage slot corresponding to the position of the mapping `balanceStates` in the storage of `BatchExchange`. This position currently happens to be the same for  `EpochTokenLocker`, so the tests should pass in the same way. If we are going to use the same tests for a different version of `BatchExchange` this could not be true anymore, and then we'd have the tests passing but the function not working as intended in practice.
I can definitely change the structure of the tests removing the setup, however.

Reply:
The inheritance layout is a good point! Up to you if you want to inline it or not (but I'd appreciate if we don't call the unnecessary `setupGenericStableX()` and instead just `await BatchExchange.new(1, erc20.address)`). 

If you leave the setup function can you please call it in a way thats explicit what is being set up (e.g. `setup_with_applied_deposit(starting_amount)`)?